### PR TITLE
8319165: hsdis binutils: warns on empty string as option string

### PR DESCRIPTION
**Pull request by robehn**
Hi, please consider.

insn_options[0] is set to empty string if there is no options (NULL or empty strings).
Checking it for empty string should cover both cases, caller option is NULL or caller option is empty string.

Tested hsdis no longer gives me the warning.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8319165](https://bugs.openjdk.org/browse/JDK-8319165): hsdis binutils: warns on empty string as option string (**Bug** - P5)

### Reviewers
 * [Hamlin Li](https://openjdk.org/census#mli) (@Hamlin-Li - **Reviewer**) ⚠️ Review applies to [94ccd7f9](https://git.openjdk.org/jdk/pull/16433/files/94ccd7f9a348c7011427860240c144f023f3b58c)
 * [Jesper Wilhelmsson](https://openjdk.org/census#jwilhelm) (@JesperIRL - **Reviewer**) ⚠️ Review applies to [94ccd7f9](https://git.openjdk.org/jdk/pull/16433/files/94ccd7f9a348c7011427860240c144f023f3b58c)
 * [Andrew Haley](https://openjdk.org/census#aph) (@theRealAph - **Reviewer**) ⚠️ Review applies to [94ccd7f9](https://git.openjdk.org/jdk/pull/16433/files/94ccd7f9a348c7011427860240c144f023f3b58c)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16433/head:pull/16433` \
`$ git checkout pull/16433`

Update a local copy of the PR: \
`$ git checkout pull/16433` \
`$ git pull https://git.openjdk.org/jdk.git pull/16433/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16433`

View PR using the GUI difftool: \
`$ git pr show -t 16433`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16433.diff">https://git.openjdk.org/jdk/pull/16433.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16433#issuecomment-1787199982)

### CODE REVIEW

File: src/utils/hsdis/binutils/hsdis-binutils.c [hunk 1]
```diff
@@ -336,8 +336,8 @@ static void setup_app_data(struct hsdis_app_data* app_data,
                                  app_data->printf_stream,
                                  app_data->printf_callback,
                                  native_bfd,
-                                 /* On PowerPC we get warnings, if we pass empty options */
-                                 (caller_options == NULL) ? NULL : app_data->insn_options);
+                                 /* On some archs we get warnings, if we pass empty options */
+                                 (app_data->insn_options[0] == '\0') ? NULL : app_data->insn_options);
```
Comment (theRealAph) [inline src/utils/hsdis/binutils/hsdis-binutils.c:340]: What if `caller_options == NULL`?

Comment (robehn) [reply src/utils/hsdis/binutils/hsdis-binutils.c:340]: The parse_caller_options handles the NULL case, so I forgot about the early bailout.
```
  for (p = caller_options; p != NULL; ) {
  }
  *iop = '\0';

```
Sorry.

Comment (robehn) [reply src/utils/hsdis/binutils/hsdis-binutils.c:340]: Just remove the if, and always call parse_caller_options ?

Comment (robehn) [reply src/utils/hsdis/binutils/hsdis-binutils.c:340]: Sorry again, long day:
```
  struct hsdis_app_data app_data;
  memset(&app_data, 0, sizeof(app_data));
```
I knew I saw that somewhere.
So if we don't parse at all (caller options NULL) it should be empty string in insn_options.

That is the way we handle caller option NULL, no ?

Comment (theRealAph) [reply src/utils/hsdis/binutils/hsdis-binutils.c:340]: Oh, I see.

[the remark this answers is not in the document]
Comment (robehn) [reply src/utils/hsdis/binutils/hsdis-binutils.c:340]: Sure, since this was not obvious.

**Review by Hamlin-Li — APPROVED** [review]
LGTM, Thanks.

**Review by JesperIRL — APPROVED** [review]
👍🏼

File: src/utils/hsdis/binutils/hsdis-binutils.c [hunk 2]
```diff
@@ -336,8 +336,9 @@ static void setup_app_data(struct hsdis_app_data* app_data,
                                  app_data->printf_stream,
                                  app_data->printf_callback,
                                  native_bfd,
-                                 /* On PowerPC we get warnings, if we pass empty options */
-                                 (caller_options == NULL) ? NULL : app_data->insn_options);
+                                 /* On some archs we get warnings, if we pass empty options */
+                                 ((caller_options == NULL) || (app_data->insn_options[0] == '\0'))
+                                 ? NULL : app_data->insn_options);
```
Comment (theRealAph) [inline src/utils/hsdis/binutils/hsdis-binutils.c:341]:
```suggestion
                                 ? NULL : app_data->insn_options);
```

Comment (theRealAph) [inline src/utils/hsdis/binutils/hsdis-binutils.c:341]:
```suggestion
                                 ((caller_options == NULL) || (app_data->insn_options[0] == '\0'))
                                   ? NULL : app_data->insn_options);
```

**Review by theRealAph — APPROVED** [review]

File: src/utils/hsdis/binutils/hsdis-binutils.c (hunk 2 above)
Comment (robehn) [reply src/utils/hsdis/binutils/hsdis-binutils.c]: Applied

**Comment by robehn** [comment]
Thanks !

**Pull request closed without merging**